src/unix/unix_internal.h

```diff
@@ -784,6 +784,13 @@ static inline void file_op_maybe_wake(thread t)
     irq_restore(flags);
 }
 
+static inline sysreturn syscall_return(thread t, sysreturn val)
+{
+    set_syscall_return(t, val);
+    file_op_maybe_wake(t);
```

Comment:
I guess you could just merge these two functions and nuke file_op_maybe_wake (there's only one use left in sendfile_bh, and that could be changed to if (bound(bh)) syscall_return(...)). Then really "file_op_is_complete" could be something more generic like "syscall_complete", and file_op_maybe_sleep could be something like "thread_maybe_sleep_uninterruptible". Then nuke file_op_begin and universally set syscall_complete to false in the syscall_debug.

In general I think it would be helpful / less confusing to unify the syscall exit points ... maybe all calls to set_syscall_return should change to syscall_return...

Reply:
OK, I nuked the file_op_* stuff and now things look more intuitive and consistent.
As for changing all calls to set_syscall_return() to calls to syscall_return(), maybe that's not ideal from an efficiency perspective, because in most cases set_syscall_return() is called by code that does a direct (non-blocking) return from a syscall, and we can avoid the overhead of disabling the irqs and all the rest. In those cases I think we should just return the syscall return value and be done with it, because set_syscall_return() is already called by syscall_debug()